src/main/java/com/streamr/client/dataunion/DataUnion.java

```diff
@@ -0,0 +1,166 @@
+import com.streamr.client.dataunion.contracts.IERC20;
+import com.streamr.client.utils.Web3jUtils;
+import org.apache.commons.lang3.reflect.FieldUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.web3j.abi.TypeEncoder;
+import org.web3j.abi.datatypes.*;
+import org.web3j.abi.datatypes.generated.Uint256;
+import org.web3j.crypto.Credentials;
+import org.web3j.crypto.ECKeyPair;
+import org.web3j.protocol.Web3j;
+import org.web3j.protocol.core.methods.response.TransactionReceipt;
+import org.web3j.tx.Contract;
+import org.web3j.tx.ManagedTransaction;
+import org.web3j.utils.Numeric;
+
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.math.BigInteger;
+import java.util.ArrayList;
+
+import static com.streamr.client.dataunion.DataUnionClient.signWithdraw;
+import static com.streamr.client.utils.Web3jUtils.waitForCodeAtAddress;
+import static com.streamr.client.utils.Web3jUtils.waitForCondition;
+
+public class DataUnion {
+    private static final Logger log = LoggerFactory.getLogger(DataUnion.class);
+    protected static final long STATUS_NONE = 0;
+    protected static final long STATUS_ACTIVE = 1;
+    protected static final long STATUS_INACTIVE = 2;
+
+    // Contract.web3j is protected, so need to access with reflection. more bad web3j design!
+    private static Web3j getWeb3j(Contract c) {
+        try {
+            return (Web3j) FieldUtils.readField(c, "web3j", true);
+        } catch (Exception e) {
+            log.error(e.getMessage());
+            throw new RuntimeException(e);
+        }
+    }
+
+    private DataUnionMainnet mainnet;
+    private DataUnionSidechain sidechain;
+
+    //use DataUnionClient to instantiate
+    protected DataUnion(DataUnionMainnet mainnet, DataUnionSidechain sidechain) {
+        this.mainnet = mainnet;
+        this.sidechain = sidechain;
+    }
+
+    public boolean waitForDeployment(long sleeptime, long timeout) throws Exception {
+        String code = waitForCodeAtAddress(sidechain.getContractAddress(), getWeb3j(sidechain), sleeptime, timeout);
+        return code != null && !code.equals("0x");
+    }
+
+    public String mainnetAddress(){
+        return mainnet.getContractAddress();
+    }
+
+    public String sidechainAddress(){
+        return sidechain.getContractAddress();
+    }
+
+    public boolean isDeployed() throws Exception {
+        //this will check the condition only once:
+        return waitForDeployment(0,0);
+    }
+
+    public void sendTokensToBridge() throws Exception {
+        mainnet.sendTokensToBridge().send();
+    }
+
+    public BigInteger waitForEarningsChange(final BigInteger initialBalance, long sleeptime, long timeout) throws Exception {
+        Web3jUtils.Condition earningsChange = new Web3jUtils.Condition(){
+            @Override
+            public Object check() throws Exception {
+                BigInteger bal = sidechain.totalEarnings().send().getValue();
+                if(!bal.equals(initialBalance))
```

Comment:
Use curly braces in if/for/etc constructs. This is a potential source for bugs in the future.